### PR TITLE
EVA-3210 - validation report

### PR DESCRIPTION
**Pull request by tcezard**
Refactor the validator as an object
Create validation parsing method 
Create category or errors in VCF validation
Create report in HTML

### CODE REVIEW

File: cli/docker_validator.py
Comment (tcezard) [inline cli/docker_validator.py:1]: These should be non functional refactor to use classes instead of functions

File: tests/test_report.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            "valid": False,
+            "warning_count": 0,
+        },
+    },
+}
+
+
+class TestReport(TestCase):
+    resource_dir = os.path.join(os.path.dirname(__file__), 'resources')
+    expected_report = os.path.join(resource_dir, 'validation_reports', 'report.html')
```
Comment (apriltuesday) [inline tests/test_report.py:57]:
```suggestion
    expected_report = os.path.join(resource_dir, 'validation_reports', 'expected_report.html')
```
Optional but I prefer having it clear which files are the inputs and which are the expected outputs for the tests.

**Review by apriltuesday — COMMENTED** [review]
I've left some suggestions but looks good overall!

File: cli/report.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+current_dir = os.path.dirname(__file__)
+
+
+def get_logo():
+    with open(os.path.join(current_dir, "eva_logo.png"), "rb") as f:
+        logo_data = base64.b64encode(f.read()).decode("utf-8")
+        return f'<img src="data:image/png;base64,{logo_data}" width="100" height="100">'
+
+
+def generate_html_report(validation_results):
```
[the remark this answers is not in the document]
Comment (tcezard) [reply cli/report.py:13]: I think you're right but at the moment I was focused on getting a first version out. I'll create a subsequent ticket.

File: cli/validator.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    def __init__(self, vcf_files, output_dir):
+        self.output_dir = output_dir
+        self.vcf_files = vcf_files
+        self.results = {}
+
+    def validate(self):
+        self._validate()
+        self._collect_validation_workflow_results()
+
+    def _validate(self):
```
[the remark this answers is not in the document]
Comment (tcezard) [reply cli/validator.py:29]: I switcher to naming it Reporter which makes more sense.

File: tests/test_docker_validator.py [hunk 4]
```diff
@@ -33,16 +32,11 @@ def setUp(self):
             for row in data:
                 writer.writerow(row)
 
-    def tearDown(self):
-        if os.path.exists(self.test_run_dir):
-            shutil.rmtree(self.test_run_dir)
```
[the remark this answers is not in the document]
Comment (tcezard) [reply tests/test_docker_validator.py:38]: Remove by mistake. Now reinstated

File: tests/test_validator.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        self.validator.create_reports()
+
+    def test_vcf_check_errors_is_critical(self):
+        errors = [
+            'INFO AC does not match the specification Number=A (expected 1 value(s)). AC=100,37.',
+            'Sample #10, field PL does not match the meta specification Number=G (expected 2 value(s)). PL=.. It must derive its number of values from the ploidy of GT (if present), or assume diploidy. Contains 1 value(s), expected 2 (derived from ploidy 1).',
+            'Sample #102, field AD does not match the meta specification Number=R (expected 3 value(s)). AD=..'
+        ]
+        for error in errors:
+            print(vcf_check_errors_is_critical(error))
```
[the remark this answers is not in the document]
Comment (tcezard) [reply tests/test_validator.py:33]: Done

File: cli/reporter.py [hunk 6]
```diff
@@ -185,6 +185,8 @@ def _collect_assembly_check_results(self):
 
     def create_reports(self):
         report_html = generate_html_report(self.results)
-        with open("report.html", "w") as f:
+        file_path = 'expected_report.html'
```
Comment (apriltuesday) [inline cli/reporter.py:188]: Sorry I didn't explain myself well in the earlier comment - this should still create a file called `report.html`, it's just that the test file used for comparison [here](https://github.com/EBIvariation/eva-sub-cli/pull/2/files/edcfbd0070862d291f770bc73a89e1f39526944d..f0c369393b6dc7ace73566aa31b8960c9c997f1c#diff-59aa45b5e4107d2c022be08538060a2f3701cfe1657eadb979ef9a077cdf8193R57) should be called `expected_report.html`, so it doesn't conflict with the actual report that the code generates, e.g. [here](https://github.com/EBIvariation/eva-sub-cli/pull/2/files/edcfbd0070862d291f770bc73a89e1f39526944d..f0c369393b6dc7ace73566aa31b8960c9c997f1c#diff-42daf86fbfce61647b461b69ee6fee5371e286ed136853ffcc88d2cfafe668b3R26).

**Review by apriltuesday — APPROVED** [review]

**Review by nitin-ebi — APPROVED** [review]

**Pull request merged**